colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -511,12 +528,17 @@ def backward_w_step(
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # loss backward; output_obj is loss
             output_obj_grad = None
-        optimizer.backward_by_grad(
-            tensor=output_obj,
-            grad=output_obj_grad,
-            inputs=list(model_chunk[model_chunk_id].parameters()),
-            retain_graph=False,
-        )
+
+        if output_obj_grad is None:
+            optimizer.backward(output_obj, inputs=list(model_chunk.parameters()), retain_graph=True)
```

Comment:
Here is the bwd_w with  dy=loss, y=output of last layer, x =input of last layer.
Should we set "retain_graph=True" at here ?

Reply:
updated.